devdoc/waiter.md

```diff
@@ -0,0 +1,71 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## Exposed API
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_OK, \
+    WAITER_RESULT_CANCELLED, \
+    WAITER_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_RESULT, WAITER_RESULT_VALUES);
+
+typedef struct WAITER_TAG* WAITER_HANDLE;
+typedef void(*NOTIFICATION_CALLBACK)(void* context, THANDLE(RC_PTR) data, WAITER_RESULT result);
```

Comment:
>THANDLE(RC_PTR) data [](http://example.com/codeflow?start=52&length=20)

If the producer already has "some" sort of THANDLE like THANDLE(RC_STRING) or THANDLE(blabla) then why would it need to have a wrapper over some void*? 

I believe producers and consumers are always going to use the same data format, and if that is already in THANDLE form, then that's what it should be passed here.

Maybe macros would help with that.

Reply:
That would require the module to be templatized on the type of the data.

I was trying to avoid macros...